Create hash_map.h

Kcchernikov opened the pull request:
Черников Кирилл, контест 3, задача A,
https://official.contest.yandex.ru/contest/24924/run-report/48698543/

Kcchernikov committed:
Create hash_map.h

Kcchernikov committed:
Update hash_map.h

Code style fixes

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+#include<algorithm>
+#include <cstddef>
```
пробел режет глаз

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+        return it != other.it;
+    }
+};
+// Hash table with open addressing. Deleted items are marked dead and cleared when the table capacity is reallocated.
```
не сказано за счёт чего достигается линейное итерирование, ничего не сказано про масштабирование, нет ссылки на подробное описание алгоритма

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+};
+// Hash table with open addressing. Deleted items are marked dead and cleared when the table capacity is reallocated.
+template<class KeyType, class ValueType, class Hash = std::hash<KeyType> >
+class HashMap {
```
это главный класс в файле, так что надо либо его поместить первым вместе с комментарием над ним, либо в заголовке файла разместить описание, что же собственно перед нами (второй предпочтительнее)

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+// Hash table with open addressing. Deleted items are marked dead and cleared when the table capacity is reallocated.
+template<class KeyType, class ValueType, class Hash = std::hash<KeyType> >
+class HashMap {
+  private:
```
приватные поля должны идти после всего остального, а не в самом начале класса

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+    char *occupied = nullptr;
+    std::vector<std::list<size_t>::iterator> iters;
+    std::list<size_t> nodes;
+    Hash hasher;
```
у приватных полей такой же способ записи (стиль оформления имён) как у публичных методов, это не ок, можно, например, добавить в конец подчёркивание

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+        nodes.clear();
+        operator delete (data);
+    }
+    // Realocate memory with new capacity/
```
почему тут в конце комментария слеш, а не точка?

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+class HashMap {
+  private:
+    size_t capacity, fullness, dead_elem;
+    std::pair<const KeyType, ValueType> *data = nullptr;
```
получается в коде смешаны низкоуровневое самостоятельное оперирование памятью и использование std::vector, что выглядит странно, почему мы не используем std::vector для data ?

Kcchernikov replied:
Видимо хотелось уменьшить время работы и место в памяти, в векторе хранится куда меньше данных чем в сырой памяти

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+         occupied = new char[capacity];         
+         init_occupied();
+    }
+    // Standard insert with open addressing, scanning from memory to the right for dead and occupied cells. Increasing the size by 10 times when the table is overloaded.
```
слишком длинный комментарий, строчка не влезает в экран

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+    void insert(const std::pair<KeyType, ValueType>& elem) {
+        size_t seed = hasher(elem.first);
+        size_t id = seed % capacity;
+        while (occupied[id] != 0 && (occupied[id] == 2 || !(data[id].first == elem.first))) {
```
тут нужен комментарий, поясняющий что происходит в этом while

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+        occupied[id] = 1;
+        nodes.push_back(id);
+        iters[id] = --nodes.end();
+        if ((fullness + dead_elem) * 10 > capacity) {
```
все управляющие константы (10 и 100 в данном случае) нужно вынести в отдельные public константы, а не держать числами в коде

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+        nodes.push_back(id);
+        iters[id] = --nodes.end();
+        if ((fullness + dead_elem) * 10 > capacity) {
+            change_capacity(std::max(static_cast<size_t>(8), fullness * 100));
```
100 это очень неэффективно

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+    void erase(const KeyType& key) {
+        size_t seed = hasher(key);
+        size_t id = seed % capacity;
+        while (occupied[id] != 0 && (occupied[id] == 2 || !(data[id].first == key))) {
```
вот этот код поиска общий с insert (и, наверняка, find), надо сделать одним методом, без копирования кода

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,383 @@
+        data = static_cast<std::pair<const KeyType, ValueType>*>(operator new(sizeof(std::pair<const KeyType, ValueType>) * capacity));
+        occupied = new char[capacity];
+        init_occupied();                  
+        auto beg2 = beg;
```
beg2 - странное название переменной

Kcchernikov committed:
Update hash_map.h

Fixed style according to requirements.

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,375 @@
+#include<memory>
+#include<list>
+
+// This is a hash table class with open addressing based on linear probing. Deleted items are marked dead and cleared when the table capacity is reallocated.
```
очень длинная строка, не влезает в экран

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,375 @@
+
+// A "bidirectional" iterator class for Hash Table formulaic of pair (key, value).
+template<class KeyType, class ValueType>
+class Literator: public std::iterator<std::forward_iterator_tag, std::pair<KeyType, ValueType>> {
```
А что означает название Literator?

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,375 @@
+        occupied_[id] = ALIVE;
+        nodes_.push_back(id);
+        iters_[id] = --nodes_.end();
+        if ((fullness_ + dead_elem_) * SIZELIMIT > capacity_) {
```
Проверка условий на то что нужно сделать масштабирование тоже должна быть отдельным методом

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,375 @@
+    }
+    // Constructor from a pair of iterators.
+    template<typename Iterator>
+    HashMap(const Iterator& beg, const Iterator& en, const Hash& _hasher_ = Hash()): fullness_(0), dead_elem_(0), hasher_(_hasher_) {
```
beg и en такие себе названия, почему не start и finish?

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,375 @@
+    // Constructor from a pair of iterators.
+    template<typename Iterator>
+    HashMap(const Iterator& beg, const Iterator& en, const Hash& _hasher_ = Hash()): fullness_(0), dead_elem_(0), hasher_(_hasher_) {
+        Iterator begin1 = beg;
```
begin1 и begin2 тоже не информативные названия

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,375 @@
+  
+   private:
+    size_t capacity_, fullness_, dead_elem_;
+    std::pair<const KeyType, ValueType> *data_ = nullptr;
```
приходи в личку вот это обсуждать, я не понимаю зачем тебе два массива с низкоуровневым выделением памяти, а дальше вектор списков

Kcchernikov committed:
Update hash_map.h

Исправил все кроме динамической памяти, по нескольким причинам, подробнее в тг.

Kcchernikov committed:
Update hash_map.h

Убрал контейнеры, оставил только сырую память

Kcchernikov committed:
Update hash_map.h

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,431 @@
+            }
+        }
+        const_iterator temp = end();
+        if (occupied_[id] != 1) return temp;
```
скобки

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,431 @@
+template<class KeyType, class ValueType, class Hash = std::hash<KeyType> >
+class HashMap {
+  public:
+    constexpr static int SIZELIMIT = 4, INCREASE = 8, MINSIZE = 8, ALIVE = 1, DEAD = 2, EMPTY = 0;
```
сделай отдельно по строке на каждую константу

glebshp replied:
и поясни что они значат

glebshp commented on hash_map.h:
```diff
@@ -0,0 +1,431 @@
+        occupied_[id] = ALIVE;
+        adapt_size_();
+    }
+    // Deleting an item by marking it dead.
```
Из-за комментариев методы сливаются, добавь пустые строки между концом метода и началом комментария к следующему

Kcchernikov committed:
Update hash_map.h
